### ISSUE TITLE
Uniqueness

### ISSUE THREAD

**Issue by jaredpalmer** [Marak/faker.js#466]
Current experiencing issues with faker not generating unique enough values. This results in a false negative test result. Our workaround is to concatenate or prepend another faker function to increase randomness.

Notoriously not-unique-enough faker funcs for running integration tests with:
- `faker.internet.ip()`
- `faker.internet.email()`

**Comment by Marak** [comment]
@jaredpalmer -

I can confirm conflicting values for both `faker.internet.ip()` and `faker.internet.email()` when running the attached code.

It seems on `master` we are seeing about 3 IP conflicts per 100,000 and 700 email conflicts per 10,000.

We should for sure fix this for email address. I should review other faker methods that may also require more entropy. RE: #245

```js

var faker = require('../../index');
faker.locale = "en";


var ips = {};
var conflicts = 0;
for (var i = 0; i < 100000; i++) {
  var ip = faker.internet.ip();
  if (typeof ips[ip] === 'undefined') {
    //console.log('set ip', ip)
    ips[ip] = ip;
  } else {
    conflicts++;
    console.log('CONFLICTS', ip)
  }
}
// 0 at 10,000
// 3 at 100,000


//console.log(faker.image.image())
var emails = {};
var conflicts = 0;
for (var i = 0; i < 100000; i++) {
  var email = faker.internet.email();
  if (typeof emails[email] === 'undefined') {
    //console.log('set email', email)
    emails[email] = email;
  } else {
    conflicts++;
    console.log('CONFLICTS', email)
  }
}
console.log('total conflicts', conflicts)

// 3 - 7 at 10,000
// 652 at 100,000
```

**Comment by Marak** [comment]
I've looked at the code for generating emails / usernames and I'm not 100% certain why we are seeing conflicts at this scale.

https://github.com/Marak/faker.js/blob/master/lib/internet.js#L69

We should be generating enough uniqueness here to not be seeing conflicts with such small datasets? Can anyone confirm the math based on the code conditionals and first / last name data?

If anything, we may have to implement a way for faker to remember previously generated results and then re-generate until a new unique is found? I feel like we can improve the randomness without having to resort to that.

**Comment by Marak** [comment]
RE: https://github.com/Marak/faker.js/pull/477

**Comment by agm1984** [comment]
I was working with an anti-duplicate script this weekend. I will see if I can produce you some logic along these lines. I'm thinking that faker should maintain some kind of local store of values and check against it to avoid duplicates. It could simply generate a new random every time it detects duplicate was already made in the session or lately (however the store is made to function).

In my testing this weekend, it is suitably performant to check against a growing text file. It took about 0.3ms to check against 1 entry and it took 10 seconds to generate 100,000 unique 16 digit values synchronously, according to console.time('foo') and console.timeEnd('foo')

The basic idea would be to create a psuedo-Redis local store, even better if it integrates with an existing Redis store.

The issue is it needs a working memory so it can avoid offering something it already offered, in my opinion. This will offer infinite horizontal scalability if a common store is used.

**Comment by Marak** [comment]
@agm1984 - Sounds good! 

Doing a pluggable Store seems like the right idea, and should be easy to do. I have been using the following `Store.js` file at scale with success. It should always work as drop-in for `node-redis` module, see: https://github.com/Stackvana/microcule/blob/master/lib/plugins/Store.js

I see the memoization / pluggable data-store as the easier part of this problem. The harder part will be generalization of code and configurability of uniqueness API. The functionality for unique API calls should probably be general enough to work on any JavaScript method? Let me know what you figure out.

There has been a bit of discussion over the years about uniqueness and tangentially related issues. If you haven't already, I'd recommend reading [this thread](https://github.com/Marak/faker.js/pull/477#issuecomment-283131302) and checking out some of the linked issues. 

Looking forward to seeing what you find. I'd also recommend deep diving into any problem faker methods you've run into in the past to see if we can make any changes to the core API itself for those methods to improve uniqueness or quality of generated data.

**Comment by agm1984** [comment]
Here is a kind of proof of concept I just made. It seems to be working perfectly for me. Feel free to use it:

https://github.com/agm1984/uniqueNameGenerator

I have to work on my main project here now lol, but I will circle back. I have a need to build a test module soon that generates fake businesses and users with a lengthy list of properties and graph DB relationships all randomizing, so I will definitely try to incorporate it into faker so I can just import faker.

I didn't want to install babel so I just rolled with this syntax. It also added 30% execution time when I used promises, so I kept it full synchronous. It was a difference of 0.3ms instead of 0.2ms per name of length 5~.

**Comment by agm1984** [comment]
The issues I had with faker, to answer your thought, were with generating Business Names, and with generating random images. I recall every method to do with images was returning the same image. It would be nice to have it pull from some free source. There should be lots of APIs for that.

If you just make a POST endpoint and a list view and dump like 10,000 records containing usage of every method, you will see where the duplication spots are :)

**Comment by Marak** [comment]
@jaredpalmer - I've done a bit of work in `master` to improve the situation for creating unique data. It's actually looking quite good.

We now have a new API method `faker.unique` to help in ensuring unique data generation from faker methods.

[Check out the example](https://github.com/Marak/faker.js/blob/master/examples/node/unique-values.js)

Looks something like this:

```js
var faker = require('../../index');

var emails = {};
var conflicts = 0;
// emails estimated: 1,055,881
// full names estimated: 1,185,139
for (var i = 0; i < 100000; i++) {

  // call function with no arguments
  var email = faker.unique(faker.internet.email);

  // or with function arguments as argument array
    // var email = faker.unique(faker.internet.email, [null, null, 'marak.com']);

  // or with custom options for maxTime as milliseconds or maxRetries
    // var email = faker.unique(faker.internet.email, [null, null, 'marak.com'], { maxRetries: 1, maxTime: 50 });

  if (typeof emails[email] === 'undefined') {
    // found a unique new item
    emails[email] = true;
  } else {
    // found a conflicting item ( should not happen using faker.unique() )
    conflicts++;
  }
}
console.log('total conflicts', conflicts); // should be zero using faker.unique()
console.log('total uniques generated', Object.keys(emails).length);
```

It should work for any faker method and will always ensure you will get a unique result. Has a few good options for configuration / safety to ensure CPU won't melt when pushing the boundaries finding unique results in set.

**Comment by Marak** [comment]
RE: #245

**Comment by Marak** [comment]
If everything looks good I'll be shipping `faker.unique` in next release. Let me know if you have any feedback.

**Comment by amackintosh** [comment]
Nice work. I love that technology direction in #245. If I have any feedback, it will be very specific. I think this will handle it, however. I look forward to next sampling, which might actually be soon since I'm working on a frontend right now.

**Comment by bonesoul** [comment]
same issue here, my tests randomly failing as faker.js can not guarantee uniqueness.

**Comment by Marak** [comment]
@bonesoul - Did you read this whole thread and or test the new functionality for uniqueness in `master` branch?

I'm not understanding your contribution to this conversation.

**Comment by bonesoul** [comment]
oh sorry didn't read it all, will check out the master.

**Comment by Marak** [comment]
Please do. It's very basic, but should be working.

Could use improvement.

**Comment by richchurcher** [comment]
@Marak the current implementation is a little buggy due to the use of `startTime` in module scope. The elapsed time gets gradually longer in comparison to this time, so it will tend to throw. Easiest way to see this is in the REPL...
```js
> const faker = require('faker')
undefined
> faker.unique(faker.internet.email)
'Crystal_Parker52@hotmail.com'
> faker.unique(faker.internet.email)
'Milo_Braun@gmail.com'
> faker.unique(faker.internet.email)
1515730819474 1515730814430 5044
found 2 unique entries before throwing error. 
retried: 0 
total time: 5044 ms
Error: exceeded maxTime for uniquness check. may not be able to generate any more unique values with current settings. try adjusting maxTime or maxRetries parameters for faker.unique()
```
Using the command history, I managed to get off two tries before it threw :smile:

**Comment by charlex** [comment]
I just ran into this in a usecase of mine, whereby we are trying to anonymize a large set of production emails. Here are the dupes (and triples) we saw in our generation:

```
COUNT    | LINE
-----------------------------------------------------
       3 | Anais47@yahoo.com
       3 | Annamae63@yahoo.com
       3 | Ardith90@yahoo.com
       3 | Breanna8@hotmail.com
       3 | Cary73@hotmail.com
       3 | Charley80@gmail.com
       3 | Colten34@gmail.com
       3 | Dave16@gmail.com
       3 | Effie58@yahoo.com
       3 | Geraldine69@hotmail.com
       3 | Jalen26@hotmail.com
       3 | Lottie42@hotmail.com
       3 | Luciano5@yahoo.com
       3 | Meda86@gmail.com
       3 | Odie56@hotmail.com
       3 | Stevie41@hotmail.com
       3 | Yessenia89@yahoo.com
       2 | Aaliyah31@hotmail.com
       2 | Abbey35@yahoo.com
       2 | Abel73@yahoo.com
       2 | Abigail27@yahoo.com
       2 | Abigale91@gmail.com
       2 | Adaline46@hotmail.com
       2 | Adam21@yahoo.com
       2 | Adelbert.Doyle@gmail.com
       2 | Adele20@gmail.com
       2 | Adelia78@gmail.com
       2 | Adelia86@yahoo.com
       2 | Adeline41@gmail.com
       2 | Adolfo51@yahoo.com
       2 | Adolfo59@hotmail.com
       2 | Adrienne68@yahoo.com
       2 | Aidan31@yahoo.com
       2 | Aidan72@gmail.com
       2 | Aimee90@gmail.com
       2 | Akeem8@hotmail.com
       2 | Al55@yahoo.com
       2 | Alana_Ward@yahoo.com
       2 | Alana3@gmail.com
       2 | Alanna19@hotmail.com
       2 | Alba43@hotmail.com
       2 | Alba45@yahoo.com
       2 | Alba91@hotmail.com
       2 | Albert_Yost@gmail.com
       2 | Albert46@yahoo.com
       2 | Albina53@gmail.com
       2 | Alden_Gerhold@hotmail.com
       2 | Alec63@yahoo.com
       2 | Alejandra86@yahoo.com
       2 | Alejandrin27@yahoo.com
       2 | Alejandrin38@gmail.com
       2 | Alex.Hirthe@gmail.com
       2 | Alexa51@yahoo.com
       2 | Alexander_Reichert@yahoo.com
       2 | Alexandria12@hotmail.com
       2 | Alexandria27@yahoo.com
       2 | Alexandro11@yahoo.com
       2 | Alexane43@gmail.com
       2 | Alexane60@yahoo.com
       2 | Alfonso.Schuppe@gmail.com
       2 | Alfonso10@gmail.com
       2 | Alfonso2@gmail.com
       2 | Alfred95@gmail.com
       2 | Alfredo89@gmail.com
       2 | Ali.Bartoletti@yahoo.com
       2 | Alice10@yahoo.com
       2 | Alicia98@yahoo.com
       2 | Alisa24@hotmail.com
       2 | Alivia_Wisozk@yahoo.com
       2 | Aliya21@yahoo.com
       2 | Alize51@gmail.com
       2 | Alize66@yahoo.com
       2 | Alize86@gmail.com
       2 | Allison6@yahoo.com
       2 | Ally51@hotmail.com
       2 | Ally82@yahoo.com
       2 | Alphonso2@hotmail.com
       2 | Althea13@yahoo.com
       2 | Alvena10@gmail.com
       2 | Alverta0@gmail.com
       2 | Alvina54@hotmail.com
       2 | Alysha97@gmail.com
       2 | Amaya0@hotmail.com
       2 | Amber90@gmail.com
       2 | Amelie73@gmail.com
       2 | Americo69@yahoo.com
       2 | Americo71@yahoo.com
       2 | Amir.Witting@yahoo.com
       2 | Amiya88@hotmail.com
       2 | Anabel50@yahoo.com
       2 | Anastasia24@gmail.com
       2 | Andreane82@hotmail.com
       2 | Andres2@hotmail.com
       2 | Andres2@yahoo.com
       2 | Andrew38@gmail.com
       2 | Andrew75@hotmail.com
       2 | Angela_Parisian@yahoo.com
       2 | Angela29@hotmail.com
       2 | Angela43@hotmail.com
       2 | Angeline48@gmail.com
       2 | Anibal29@gmail.com
       2 | Anne62@hotmail.com
       2 | Annetta.Russel@hotmail.com
       2 | Annetta75@gmail.com
       2 | Annie85@gmail.com
       2 | Ansel21@yahoo.com
       2 | Antwon82@gmail.com
       2 | April73@yahoo.com
       2 | Araceli84@gmail.com
       2 | Arely90@yahoo.com
       2 | Aric63@yahoo.com
       2 | Ariel11@hotmail.com
       2 | Arlie31@gmail.com
       2 | Arlo_Hirthe@gmail.com
       2 | Arlo82@hotmail.com
       2 | Armani9@yahoo.com
       2 | Arne26@gmail.com
       2 | Asha73@gmail.com
       2 | Ashleigh55@gmail.com
       2 | Ashly18@gmail.com
       2 | Aubrey23@yahoo.com
       2 | Audra_Ferry@hotmail.com
       2 | Audra53@hotmail.com
       2 | Audra68@yahoo.com
       2 | Audreanne69@yahoo.com
       2 | Austin.Carroll@yahoo.com
       2 | Avery94@gmail.com
       2 | Axel62@hotmail.com
       2 | Ayana70@hotmail.com
       2 | Ayden81@gmail.com
       2 | Barry23@gmail.com
       2 | Barton15@hotmail.com
       2 | Barton88@gmail.com
       2 | Baylee.Corkery@gmail.com
       2 | Beatrice33@gmail.com
       2 | Beaulah2@hotmail.com
       2 | Benedict23@gmail.com
       2 | Benjamin63@gmail.com
       2 | Bennett32@yahoo.com
       2 | Berenice55@yahoo.com
       2 | Bernadette71@gmail.com
       2 | Bernadine35@gmail.com
       2 | Bernadine4@gmail.com
       2 | Bernardo23@hotmail.com
       2 | Bernie.Dickens@yahoo.com
       2 | Bertha28@yahoo.com
       2 | Beryl45@yahoo.com
       2 | Bethel10@yahoo.com
       2 | Bettie40@hotmail.com
       2 | Bettye_Hauck@yahoo.com
       2 | Beulah75@gmail.com
       2 | Beverly5@hotmail.com
       2 | Bill72@gmail.com
       2 | Billie85@hotmail.com
       2 | Blaise17@gmail.com
       2 | Blaise61@hotmail.com
       2 | Blake52@gmail.com
       2 | Bobby42@gmail.com
       2 | Boyd82@gmail.com
       2 | Bradford23@hotmail.com
       2 | Bradford36@gmail.com
       2 | Bradly_Heathcote@yahoo.com
       2 | Braeden42@yahoo.com
       2 | Braeden52@yahoo.com
       2 | Brain20@yahoo.com
       2 | Brandi_Parisian@hotmail.com
       2 | Brandyn17@yahoo.com
       2 | Brant83@gmail.com
       2 | Breana40@gmail.com
       2 | Brent_Jerde@gmail.com
       2 | Brent28@gmail.com
       2 | Brett.Daugherty@yahoo.com
       2 | Briana45@hotmail.com
       2 | Brittany47@gmail.com
       2 | Brock62@yahoo.com
       2 | Brook56@gmail.com
       2 | Brooke13@hotmail.com
       2 | Bryana40@yahoo.com
       2 | Buck55@gmail.com
       2 | Burdette.Russel@hotmail.com
       2 | Burdette58@gmail.com
       2 | Buster_Schneider@yahoo.com
       2 | Cali75@yahoo.com
       2 | Callie72@hotmail.com
       2 | Camila26@yahoo.com
       2 | Camilla62@yahoo.com
       2 | Candace35@hotmail.com
       2 | Candace9@hotmail.com
       2 | Carlee30@yahoo.com
       2 | Carli14@yahoo.com
       2 | Carli42@hotmail.com
       2 | Carmella23@yahoo.com
       2 | Carmelo76@gmail.com
       2 | Carolyn41@hotmail.com
       2 | Carter92@hotmail.com
       2 | Cassandra_Green@gmail.com
       2 | Cassandra.Jacobson@hotmail.com
       2 | Cassandra81@hotmail.com
       2 | Catalina26@yahoo.com
       2 | Catherine28@yahoo.com
       2 | Catherine45@yahoo.com
       2 | Cathryn22@hotmail.com
       2 | Cathy35@gmail.com
       2 | Cecil34@hotmail.com
       2 | Cedrick75@hotmail.com
       2 | Chadd46@gmail.com
       2 | Chanelle39@hotmail.com
       2 | Chelsey34@gmail.com
       2 | Chelsie50@hotmail.com
       2 | Christiana33@yahoo.com
       2 | Christophe87@hotmail.com
       2 | Christopher19@gmail.com
       2 | Christopher26@hotmail.com
       2 | Cicero2@hotmail.com
       2 | Clarabelle78@gmail.com
       2 | Clark13@hotmail.com
       2 | Claude75@yahoo.com
       2 | Claudine22@hotmail.com
       2 | Clay54@hotmail.com
       2 | Clementine95@hotmail.com
       2 | Cletus16@gmail.com
       2 | Cleve35@gmail.com
       2 | Clifton4@gmail.com
       2 | Clinton25@gmail.com
       2 | Clinton73@yahoo.com
       2 | Clotilde_Cormier@hotmail.com
       2 | Clyde75@hotmail.com
       2 | Coby92@hotmail.com
       2 | Cody11@yahoo.com
       2 | Cole41@gmail.com
       2 | Coleman84@gmail.com
       2 | Connie20@hotmail.com
       2 | Connor48@gmail.com
       2 | Conrad97@hotmail.com
       2 | Consuelo52@hotmail.com
       2 | Cooper22@yahoo.com
       2 | Cora86@gmail.com
       2 | Coralie71@gmail.com
       2 | Corene90@hotmail.com
       2 | Corine19@hotmail.com
       2 | Cornell30@gmail.com
       2 | Corrine60@hotmail.com
       2 | Courtney82@gmail.com
       2 | Coy_Grant@hotmail.com
       2 | Crawford41@gmail.com
       2 | Cristal.Blick@yahoo.com
       2 | Cruz3@yahoo.com
       2 | Curtis66@yahoo.com
       2 | Curtis98@hotmail.com
       2 | Dahlia3@yahoo.com
       2 | Daija74@gmail.com
       2 | Daisha67@gmail.com
       2 | Dallas92@yahoo.com
       2 | Damion27@hotmail.com
       2 | Damion75@yahoo.com
       2 | Dana55@yahoo.com
       2 | Dangelo45@gmail.com
       2 | Dangelo9@yahoo.com
       2 | Dangelo95@hotmail.com
       2 | Dangelo97@gmail.com
       2 | Danial96@hotmail.com
       2 | Danielle2@hotmail.com
       2 | Danielle6@hotmail.com
       2 | Danika40@hotmail.com
       2 | Dante58@hotmail.com
       2 | Danyka27@hotmail.com
       2 | Daphne35@yahoo.com
       2 | Darby13@hotmail.com
       2 | Darian42@gmail.com
       2 | Darian93@gmail.com
       2 | Dario57@yahoo.com
       2 | Darren12@gmail.com
       2 | Darron90@hotmail.com
       2 | Davin54@hotmail.com
       2 | Davion38@hotmail.com
       2 | Deangelo39@yahoo.com
       2 | Deangelo77@hotmail.com
       2 | Deanna70@yahoo.com
       2 | Deborah.Hyatt@yahoo.com
       2 | Dejah55@hotmail.com
       2 | Delaney_Olson@hotmail.com
       2 | Delfina55@yahoo.com
       2 | Delmer32@yahoo.com
       2 | Delpha22@yahoo.com
       2 | Delphine18@hotmail.com
       2 | Demarco35@gmail.com
       2 | Demario63@hotmail.com
       2 | Demetris3@hotmail.com
       2 | Demetrius4@gmail.com
       2 | Demetrius5@hotmail.com
       2 | Demetrius64@gmail.com
       2 | Dena_Kessler@hotmail.com
       2 | Deshaun78@gmail.com
       2 | Deshawn97@gmail.com
       2 | Desiree3@hotmail.com
       2 | Dessie11@gmail.com
       2 | Dessie33@yahoo.com
       2 | Dessie47@yahoo.com
       2 | Destany38@gmail.com
       2 | Destin80@hotmail.com
       2 | Devante15@hotmail.com
       2 | Devon23@hotmail.com
       2 | Dillan.Langosh@yahoo.com
       2 | Dimitri85@yahoo.com
       2 | Dina39@yahoo.com
       2 | Dino90@yahoo.com
       2 | Dolly_Heaney@hotmail.com
       2 | Domenic48@yahoo.com
       2 | Domenick7@yahoo.com
       2 | Domenico0@hotmail.com
       2 | Dominique47@gmail.com
       2 | Donald35@gmail.com
       2 | Donnell94@hotmail.com
       2 | Dorian47@gmail.com
       2 | Dorris46@hotmail.com
       2 | Dortha_Walker@yahoo.com
       2 | Dortha95@gmail.com
       2 | Dortha99@hotmail.com
       2 | Dorthy.Zemlak@hotmail.com
       2 | Douglas50@yahoo.com
       2 | Doyle66@hotmail.com
       2 | Drew85@gmail.com
       2 | Dustin67@hotmail.com
       2 | Dwight8@hotmail.com
       2 | Dwight83@gmail.com
       2 | Earl_Pagac@yahoo.com
       2 | Earl94@yahoo.com
       2 | Earline62@gmail.com
       2 | Earnestine72@gmail.com
       2 | Easter64@gmail.com
       2 | Easter82@hotmail.com
       2 | Ebba30@gmail.com
       2 | Edd.Douglas@gmail.com
       2 | Edd78@hotmail.com
       2 | Edgar25@gmail.com
       2 | Edmond8@yahoo.com
       2 | Edmund66@yahoo.com
       2 | Eduardo15@hotmail.com
       2 | Edythe53@yahoo.com
       2 | Efren85@yahoo.com
       2 | Elbert80@yahoo.com
       2 | Eldred55@yahoo.com
       2 | Electa77@gmail.com
       2 | Electa94@gmail.com
       2 | Elenora.Hand@hotmail.com
       2 | Eliezer55@hotmail.com
       2 | Elijah40@gmail.com
       2 | Elijah57@yahoo.com
       2 | Elisha42@yahoo.com
       2 | Elissa.Corkery@gmail.com
       2 | Eliza18@hotmail.com
       2 | Ella73@gmail.com
       2 | Ella88@gmail.com
       2 | Elliott88@yahoo.com
       2 | Ellis1@yahoo.com
       2 | Ellis86@hotmail.com
       2 | Elmira.Lind@yahoo.com
       2 | Elmo56@hotmail.com
       2 | Elmore87@yahoo.com
       2 | Elouise_Ziemann@hotmail.com
       2 | Elouise9@gmail.com
       2 | Elroy38@hotmail.com
       2 | Elsa5@gmail.com
       2 | Elta23@yahoo.com
       2 | Elton95@gmail.com
       2 | Elyse74@hotmail.com
       2 | Elza52@gmail.com
       2 | Emely31@gmail.com
       2 | Emerald.Roob@hotmail.com
       2 | Emerson32@hotmail.com
       2 | Emil47@gmail.com
       2 | Emil84@gmail.com
       2 | Emile55@gmail.com
       2 | Emilio78@hotmail.com
       2 | Emmanuel99@yahoo.com
       2 | Emmitt_Will@yahoo.com
       2 | Emmy34@gmail.com
       2 | Emmy83@gmail.com
       2 | Enid20@yahoo.com
       2 | Enid8@gmail.com
       2 | Enoch84@hotmail.com
       2 | Enrique39@hotmail.com
       2 | Eriberto97@gmail.com
       2 | Erich24@hotmail.com
       2 | Erik51@gmail.com
       2 | Erin38@gmail.com
       2 | Erna3@gmail.com
       2 | Ernest20@yahoo.com
       2 | Ernest40@gmail.com
       2 | Erwin3@gmail.com
       2 | Esteban60@gmail.com
       2 | Estefania17@hotmail.com
       2 | Estefania86@yahoo.com
       2 | Estella39@gmail.com
       2 | Estelle47@hotmail.com
       2 | Ethyl0@hotmail.com
       2 | Eugenia16@yahoo.com
       2 | Eulalia54@gmail.com
       2 | Evangeline56@gmail.com
       2 | Everardo.Beer@gmail.com
       2 | Ewald77@yahoo.com
       2 | Ezequiel32@gmail.com
       2 | Ezra37@hotmail.com
       2 | Fabian47@gmail.com
       2 | Fabiola76@gmail.com
       2 | Fausto42@hotmail.com
       2 | Fausto9@hotmail.com
       2 | Favian85@hotmail.com
       2 | Federico73@hotmail.com
       2 | Felicity67@hotmail.com
       2 | Felicity87@hotmail.com
       2 | Felton78@hotmail.com
       2 | Felton96@gmail.com
       2 | Fernando53@hotmail.com
       2 | Fidel79@gmail.com
       2 | Fiona41@hotmail.com
       2 | Flavie61@gmail.com
       2 | Flavio31@yahoo.com
       2 | Florencio80@gmail.com
       2 | Florida94@yahoo.com
       2 | Flossie36@hotmail.com
       2 | Floyd45@yahoo.com
       2 | Foster7@gmail.com
       2 | Frank72@yahoo.com
       2 | Frederic15@yahoo.com
       2 | Frederick57@hotmail.com
       2 | Fredy14@gmail.com
       2 | Fredy42@yahoo.com
       2 | Friedrich69@yahoo.com
       2 | Furman12@gmail.com
       2 | Gabriella.Hartmann@hotmail.com
       2 | Gabriella14@hotmail.com
       2 | Gaetano10@gmail.com
       2 | Gaetano74@hotmail.com
       2 | Gage10@hotmail.com
       2 | Garett96@hotmail.com
       2 | Garland19@hotmail.com
       2 | Garrett8@gmail.com
       2 | Garrick6@yahoo.com
       2 | Gayle59@hotmail.com
       2 | Gene59@hotmail.com
       2 | General55@gmail.com
       2 | Geo27@yahoo.com
       2 | Geoffrey53@gmail.com
       2 | Georgette91@hotmail.com
       2 | Geovany36@gmail.com
       2 | Gerald84@gmail.com
       2 | Gerardo74@hotmail.com
       2 | Gerda63@gmail.com
       2 | German89@hotmail.com
       2 | Gerry39@hotmail.com
       2 | Gertrude.Davis@hotmail.com
       2 | Gertrude.Metz@gmail.com
       2 | Gertrude25@hotmail.com
       2 | Gideon15@yahoo.com
       2 | Gilda33@yahoo.com
       2 | Giovanni56@yahoo.com
       2 | Giovanny16@yahoo.com
       2 | Giuseppe35@hotmail.com
       2 | Glen97@gmail.com
       2 | Glenda28@gmail.com
       2 | Glenda54@yahoo.com
       2 | Graham46@hotmail.com
       2 | Grant70@hotmail.com
       2 | Gregoria74@hotmail.com
       2 | Greyson89@yahoo.com
       2 | Grover91@yahoo.com
       2 | Guadalupe0@yahoo.com
       2 | Gudrun.Bahringer@gmail.com
       2 | Gudrun37@hotmail.com
       2 | Guido5@yahoo.com
       2 | Guiseppe13@gmail.com
       2 | Gunner6@hotmail.com
       2 | Gus52@yahoo.com
       2 | Gust58@yahoo.com
       2 | Gwendolyn93@gmail.com
       2 | Hailey72@hotmail.com
       2 | Hanna22@hotmail.com
       2 | Hanna39@yahoo.com
       2 | Harley51@hotmail.com
       2 | Harmon20@yahoo.com
       2 | Hassan_Herman@gmail.com
       2 | Hattie77@hotmail.com
       2 | Haylie3@gmail.com
       2 | Heath.Gerlach@hotmail.com
       2 | Heather21@gmail.com
       2 | Helene15@hotmail.com
       2 | Heloise13@gmail.com
       2 | Henderson3@hotmail.com
       2 | Henri_Schinner@hotmail.com
       2 | Herminia42@yahoo.com
       2 | Herminio37@gmail.com
       2 | Hershel_Conroy@gmail.com
       2 | Herta.Quitzon@hotmail.com
       2 | Hettie.Conn@gmail.com
       2 | Hillard_Rath@yahoo.com
       2 | Hillard85@gmail.com
       2 | Hillary18@gmail.com
       2 | Hilma87@gmail.com
       2 | Hilton40@yahoo.com
       2 | Hipolito72@yahoo.com
       2 | Hiram82@gmail.com
       2 | Hobart51@hotmail.com
       2 | Hollie21@gmail.com
       2 | Hoyt54@yahoo.com
       2 | Hubert0@yahoo.com
       2 | Hugh57@gmail.com
       2 | Humberto74@yahoo.com
       2 | Icie56@yahoo.com
       2 | Idella_Padberg@hotmail.com
       2 | Idella25@hotmail.com
       2 | Ignacio56@yahoo.com
       2 | Ignacio72@hotmail.com
       2 | Ignacio97@gmail.com
       2 | Ignatius25@gmail.com
       2 | Imani.Waelchi@yahoo.com
       2 | Imelda20@hotmail.com
       2 | Irma22@yahoo.com
       2 | Irma65@gmail.com
       2 | Irving53@yahoo.com
       2 | Isaac94@yahoo.com
       2 | Isabel85@yahoo.com
       2 | Isabell76@hotmail.com
       2 | Isabell77@gmail.com
       2 | Isabella13@gmail.com
       2 | Isac35@hotmail.com
       2 | Isadore73@gmail.com
       2 | Isaiah28@yahoo.com
       2 | Isaiah30@yahoo.com
       2 | Ismael47@hotmail.com
       2 | Isobel68@gmail.com
       2 | Itzel_Brekke@yahoo.com
       2 | Iva98@hotmail.com
       2 | Jackeline17@gmail.com
       2 | Jackie95@hotmail.com
       2 | Jacques61@gmail.com
       2 | Jaeden81@gmail.com
       2 | Jaida6@gmail.com
       2 | Jaiden83@yahoo.com
       2 | Jakayla40@yahoo.com
       2 | Jaleel12@gmail.com
       2 | Jalon90@gmail.com
       2 | Jamel7@hotmail.com
       2 | Jan33@gmail.com
       2 | Janessa45@gmail.com
       2 | Janessa68@hotmail.com
       2 | Janick94@yahoo.com
       2 | Janis49@hotmail.com
       2 | Janiya59@gmail.com
       2 | Jannie69@hotmail.com
       2 | Jany10@gmail.com
       2 | Jaquan82@yahoo.com
       2 | Jaren36@hotmail.com
       2 | Jaren37@yahoo.com
       2 | Jarrell20@yahoo.com
       2 | Jarvis51@yahoo.com
       2 | Jasper0@gmail.com
       2 | Jaunita5@hotmail.com
       2 | Jaylan20@hotmail.com
       2 | Jaylen_Bechtelar@hotmail.com
       2 | Jaylon0@yahoo.com
       2 | Jayne11@hotmail.com
       2 | Jayne44@yahoo.com
       2 | Jean18@gmail.com
       2 | Jeanne67@gmail.com
       2 | Jed72@gmail.com
       2 | Jeffery61@hotmail.com
       2 | Jeffery81@hotmail.com
       2 | Jeffrey.Corwin@gmail.com
       2 | Jeffry76@hotmail.com
       2 | Jeffry86@gmail.com
       2 | Jenifer_Gorczany@gmail.com
       2 | Jennings29@gmail.com
       2 | Jennyfer28@hotmail.com
       2 | Jeramie33@yahoo.com
       2 | Jeramie37@hotmail.com
       2 | Jerrell43@yahoo.com
       2 | Jerry54@hotmail.com
       2 | Jesse69@gmail.com
       2 | Jessica60@hotmail.com
       2 | Jessie98@hotmail.com
       2 | Jessika62@yahoo.com
       2 | Jessika99@hotmail.com
       2 | Jessy30@gmail.com
       2 | Jessy38@gmail.com
       2 | Jevon30@gmail.com
       2 | Jewell20@yahoo.com
       2 | Jimmy85@hotmail.com
       2 | Joan80@yahoo.com
       2 | Joanny19@gmail.com
       2 | Joaquin35@gmail.com
       2 | Joelle45@gmail.com
       2 | Joelle73@yahoo.com
       2 | Joesph21@gmail.com
       2 | Johann13@yahoo.com
       2 | Johann77@gmail.com
       2 | Johanna_Reynolds@hotmail.com
       2 | Johathan88@gmail.com
       2 | Johnpaul71@yahoo.com
       2 | Jon.Macejkovic@gmail.com
       2 | Jonas5@yahoo.com
       2 | Jonas81@yahoo.com
       2 | Jonathon79@yahoo.com
       2 | Jordan91@gmail.com
       2 | Jordon44@hotmail.com
       2 | Jorge81@hotmail.com
       2 | Joseph48@yahoo.com
       2 | Jovani78@yahoo.com
       2 | Jovanny48@gmail.com
       2 | Jovanny90@hotmail.com
       2 | Jude_Goldner@gmail.com
       2 | Judge66@hotmail.com
       2 | Judson.Leuschke@gmail.com
       2 | Juliet_Schoen@yahoo.com
       2 | Juliet53@hotmail.com
       2 | Julius5@hotmail.com
       2 | Justus1@hotmail.com
       2 | Justus72@gmail.com
       2 | Kacey77@gmail.com
       2 | Kaleigh88@hotmail.com
       2 | Kali73@hotmail.com
       2 | Kameron66@yahoo.com
       2 | Kamron66@hotmail.com
       2 | Karen17@gmail.com
       2 | Kari.Okuneva@hotmail.com
       2 | Karlee54@yahoo.com
       2 | Karson_Altenwerth@gmail.com
       2 | Kasandra33@yahoo.com
       2 | Katarina80@gmail.com
       2 | Katelynn19@hotmail.com
       2 | Katelynn91@yahoo.com
       2 | Katharina3@hotmail.com
       2 | Katherine16@yahoo.com
       2 | Katheryn_Kulas@gmail.com
       2 | Katheryn96@yahoo.com
       2 | Kathlyn41@yahoo.com
       2 | Kathryne99@yahoo.com
       2 | Katlynn62@hotmail.com
       2 | Kattie96@hotmail.com
       2 | Kavon63@yahoo.com
       2 | Kavon83@gmail.com
       2 | Kaylee5@yahoo.com
       2 | Kayli78@gmail.com
       2 | Keanu86@hotmail.com
       2 | Keara76@yahoo.com
       2 | Keegan18@hotmail.com
       2 | Keegan2@gmail.com
       2 | Keeley56@gmail.com
       2 | Kelley88@gmail.com
       2 | Kelsie59@gmail.com
       2 | Kendall.Barton@hotmail.com
       2 | Kennedy53@yahoo.com
       2 | Kenneth32@hotmail.com
       2 | Keshaun94@gmail.com
       2 | Keven23@hotmail.com
       2 | Keyon46@hotmail.com
       2 | Keyshawn44@hotmail.com
       2 | Khalil62@yahoo.com
       2 | Kiarra24@yahoo.com
       2 | Kiel83@hotmail.com
       2 | Kiera17@hotmail.com
       2 | Kim75@gmail.com
       2 | Kira56@yahoo.com
       2 | Kirstin30@gmail.com
       2 | Kirstin58@hotmail.com
       2 | Kody48@hotmail.com
       2 | Kody66@hotmail.com
       2 | Kolby91@gmail.com
       2 | Kristina5@hotmail.com
       2 | Kristoffer_Vandervort@gmail.com
       2 | Kristoffer60@hotmail.com
       2 | Kristopher6@hotmail.com
       2 | Kristy55@yahoo.com
       2 | Kyler50@gmail.com
       2 | Kylie22@yahoo.com
       2 | Kyra51@hotmail.com
       2 | Kyra72@gmail.com
       2 | Lacey82@hotmail.com
       2 | Lacy3@gmail.com
       2 | Ladarius39@gmail.com
       2 | Lafayette.Marvin@gmail.com
       2 | Laila1@gmail.com
       2 | Laisha38@gmail.com
       2 | Lambert39@hotmail.com
       2 | Lamont19@gmail.com
       2 | Landen23@hotmail.com
       2 | Larissa_Koepp@gmail.com
       2 | Laurianne25@gmail.com
       2 | Lauryn85@hotmail.com
       2 | Laverna38@gmail.com
       2 | Laverne12@yahoo.com
       2 | Laverne48@hotmail.com
       2 | Lavon89@gmail.com
       2 | Lea53@yahoo.com
       2 | Leann42@gmail.com
       2 | Leanna43@hotmail.com
       2 | Lenny26@hotmail.com
       2 | Lenore90@gmail.com
       2 | Lenore90@hotmail.com
       2 | Leonard_Braun@gmail.com
       2 | Leonel.Collins@hotmail.com
       2 | Leonel34@yahoo.com
       2 | Leonor.Smith@gmail.com
       2 | Leonor57@gmail.com
       2 | Leopold12@hotmail.com
       2 | Leopoldo50@gmail.com
       2 | Lester48@yahoo.com
       2 | Leta84@hotmail.com
       2 | Letitia41@gmail.com
       2 | Lewis52@hotmail.com
       2 | Lewis54@yahoo.com
       2 | Liam47@hotmail.com
       2 | Liam80@yahoo.com
       2 | Lila31@hotmail.com
       2 | Liliana27@gmail.com
       2 | Liliana70@gmail.com
       2 | Liliane39@yahoo.com
       2 | Lilly79@yahoo.com
       2 | Lilly85@gmail.com
       2 | Lincoln4@yahoo.com
       2 | Linda97@yahoo.com
       2 | Lindsey29@gmail.com
       2 | Linnea52@gmail.com
       2 | Linnie33@gmail.com
       2 | Lionel53@yahoo.com
       2 | Lisandro2@gmail.com
       2 | Litzy19@hotmail.com
       2 | Llewellyn20@hotmail.com
       2 | Lloyd92@hotmail.com
       2 | Lois17@yahoo.com
       2 | Lois40@hotmail.com
       2 | Loma17@hotmail.com
       2 | London96@gmail.com
       2 | Lonie21@hotmail.com
       2 | Lonie45@gmail.com
       2 | Lonzo43@yahoo.com
       2 | Lorine79@gmail.com
       2 | Lorna12@yahoo.com
       2 | Lorna81@hotmail.com
       2 | Loyal47@yahoo.com
       2 | Lucienne0@gmail.com
       2 | Lucienne80@yahoo.com
       2 | Lucio96@hotmail.com
       2 | Lucy25@gmail.com
       2 | Luis56@gmail.com
       2 | Luisa80@yahoo.com
       2 | Luna29@hotmail.com
       2 | Lyric3@hotmail.com
       2 | Lyric62@yahoo.com
       2 | Lysanne42@hotmail.com
       2 | Mac53@hotmail.com
       2 | Mac85@yahoo.com
       2 | Maci37@hotmail.com
       2 | Mack58@yahoo.com
       2 | Mack9@hotmail.com
       2 | Madeline_Monahan@gmail.com
       2 | Madeline.Ferry@gmail.com
       2 | Madelyn94@hotmail.com
       2 | Madison14@hotmail.com
       2 | Madisyn65@yahoo.com
       2 | Madonna_Kassulke@yahoo.com
       2 | Madonna82@gmail.com
       2 | Maegan67@gmail.com
       2 | Maia10@gmail.com
       2 | Makayla.White@hotmail.com
       2 | Mallory41@hotmail.com
       2 | Mara.Gottlieb@gmail.com
       2 | Marcelina40@yahoo.com
       2 | Marcelle84@yahoo.com
       2 | Marcelo66@yahoo.com
       2 | Marco_Kutch@gmail.com
       2 | Margaret_McLaughlin@yahoo.com
       2 | Margaret60@yahoo.com
       2 | Margaret69@yahoo.com
       2 | Margarete15@gmail.com
       2 | Margarete40@hotmail.com
       2 | Margarete59@gmail.com
       2 | Margie43@gmail.com
       2 | Margret13@gmail.com
       2 | Margret55@yahoo.com
       2 | Maria74@hotmail.com
       2 | Mariah49@hotmail.com
       2 | Mariana36@hotmail.com
       2 | Marianna3@yahoo.com
       2 | Marianne4@gmail.com
       2 | Marianne88@gmail.com
       2 | Marie0@gmail.com
       2 | Marietta84@hotmail.com
       2 | Marilie44@yahoo.com
       2 | Marilou19@yahoo.com
       2 | Marilou35@yahoo.com
       2 | Mario92@hotmail.com
       2 | Markus12@gmail.com
       2 | Marlene.Ankunding@yahoo.com
       2 | Marley25@hotmail.com
       2 | Marques58@gmail.com
       2 | Marques58@yahoo.com
       2 | Marta14@yahoo.com
       2 | Marta17@gmail.com
       2 | Martine71@hotmail.com
       2 | Maryam11@hotmail.com
       2 | Maud.Conroy@gmail.com
       2 | Maudie2@gmail.com
       2 | Maurice95@hotmail.com
       2 | Max.Kreiger@yahoo.com
       2 | Meda71@yahoo.com
       2 | Melany30@gmail.com
       2 | Mellie3@hotmail.com
       2 | Mellie43@hotmail.com
       2 | Melyssa45@yahoo.com
       2 | Meredith19@gmail.com
       2 | Meredith54@gmail.com
       2 | Meredith56@gmail.com
       2 | Merle13@hotmail.com
       2 | Merle8@yahoo.com
       2 | Mertie2@hotmail.com
       2 | Meta64@yahoo.com
       2 | Mia10@gmail.com
       2 | Mia49@hotmail.com
       2 | Michael7@hotmail.com
       2 | Micheal36@yahoo.com
       2 | Michel98@gmail.com
       2 | Mike66@yahoo.com
       2 | Milan36@gmail.com
       2 | Milan5@yahoo.com
       2 | Miles19@hotmail.com
       2 | Millie49@yahoo.com
       2 | Milton11@gmail.com
       2 | Missouri14@yahoo.com
       2 | Misty97@yahoo.com
       2 | Mitchel99@hotmail.com
       2 | Molly27@gmail.com
       2 | Monroe_Vandervort@gmail.com
       2 | Monserrat78@yahoo.com
       2 | Morgan_OKeefe@yahoo.com
       2 | Moses60@yahoo.com
       2 | Murphy8@yahoo.com
       2 | Myrl11@yahoo.com
       2 | Myrna48@yahoo.com
       2 | Nadia71@gmail.com
       2 | Nadia86@yahoo.com
       2 | Nakia63@gmail.com
       2 | Nakia81@yahoo.com
       2 | Name9@yahoo.com
       2 | Nannie78@hotmail.com
       2 | Naomi53@hotmail.com
       2 | Nat53@hotmail.com
       2 | Natalie85@hotmail.com
       2 | Natasha.Kerluke@yahoo.com
       2 | Nathan14@gmail.com
       2 | Nathanael33@hotmail.com
       2 | Nathanial.Turner@yahoo.com
       2 | Nathanial71@gmail.com
       2 | Ned88@hotmail.com
       2 | Neil30@hotmail.com
       2 | Nelda_Torphy@yahoo.com
       2 | Nestor62@hotmail.com
       2 | Nettie12@gmail.com
       2 | Nettie2@hotmail.com
       2 | Neva52@yahoo.com
       2 | Nichole33@yahoo.com
       2 | Nikki80@hotmail.com
       2 | Nils49@gmail.com
       2 | Noah23@yahoo.com
       2 | Noe58@gmail.com
       2 | Norma_Kuhn@hotmail.com
       2 | Norwood33@gmail.com
       2 | Nova43@yahoo.com
       2 | Nyasia78@yahoo.com
       2 | Obie72@gmail.com
       2 | Oda2@hotmail.com
       2 | Odie22@hotmail.com
       2 | Oleta20@hotmail.com
       2 | Oliver23@yahoo.com
       2 | Omer94@yahoo.com
       2 | Opal52@yahoo.com
       2 | Ora38@hotmail.com
       2 | Oran70@hotmail.com
       2 | Orie_Osinski@gmail.com
       2 | Orval1@gmail.com
       2 | Orville89@yahoo.com
       2 | Osvaldo18@yahoo.com
       2 | Osvaldo97@yahoo.com
       2 | Otilia69@hotmail.com
       2 | Otto17@yahoo.com
       2 | Ova51@gmail.com
       2 | Ova68@hotmail.com
       2 | Ozella27@hotmail.com
       2 | Pablo8@gmail.com
       2 | Pamela31@hotmail.com
       2 | Pansy62@hotmail.com
       2 | Pascale97@hotmail.com
       2 | Pat22@gmail.com
       2 | Patience41@hotmail.com
       2 | Patience47@hotmail.com
       2 | Patsy43@hotmail.com
       2 | Paul43@yahoo.com
       2 | Paula36@hotmail.com
       2 | Paxton68@hotmail.com
       2 | Pearlie49@hotmail.com
       2 | Pedro6@yahoo.com
       2 | Peggie79@hotmail.com
       2 | Percy19@hotmail.com
       2 | Percy4@gmail.com
       2 | Peyton24@gmail.com
       2 | Philip81@hotmail.com
       2 | Phoebe63@gmail.com
       2 | Piper3@hotmail.com
       2 | Presley32@hotmail.com
       2 | Queen31@gmail.com
       2 | Quinten.Haag@hotmail.com
       2 | Rae33@yahoo.com
       2 | Raleigh3@hotmail.com
       2 | Raleigh46@gmail.com
       2 | Ralph31@yahoo.com
       2 | Randall25@gmail.com
       2 | Raquel15@hotmail.com
       2 | Rashad_Mohr@gmail.com
       2 | Raymond8@hotmail.com
       2 | Raymundo49@hotmail.com
       2 | Raymundo67@gmail.com
       2 | Reagan22@gmail.com
       2 | Rebeka1@yahoo.com
       2 | Rebekah91@gmail.com
       2 | Retha32@hotmail.com
       2 | Reuben33@hotmail.com
       2 | Reuben41@gmail.com
       2 | Reyes79@hotmail.com
       2 | Reymundo17@yahoo.com
       2 | Rhett8@yahoo.com
       2 | Rhett84@yahoo.com
       2 | Rhianna28@hotmail.com
       2 | Richard15@yahoo.com
       2 | Richmond27@yahoo.com
       2 | Riley24@gmail.com
       2 | River60@yahoo.com
       2 | Rodrigo18@yahoo.com
       2 | Roel9@gmail.com
       2 | Rogers65@yahoo.com
       2 | Rolando5@gmail.com
       2 | Rolando65@gmail.com
       2 | Romaine72@yahoo.com
       2 | Roman68@hotmail.com
       2 | Roman98@hotmail.com
       2 | Ron13@hotmail.com
       2 | Ronaldo_Christiansen@gmail.com
       2 | Ronaldo12@yahoo.com
       2 | Rosa_Larson@hotmail.com
       2 | Rosa38@hotmail.com
       2 | Rosalia79@gmail.com
       2 | Rosalind87@hotmail.com
       2 | Rosalinda70@yahoo.com
       2 | Rosalinda81@hotmail.com
       2 | Rosamond83@gmail.com
       2 | Roslyn39@yahoo.com
       2 | Rossie6@gmail.com
       2 | Roxane59@yahoo.com
       2 | Roxanne92@hotmail.com
       2 | Roy28@gmail.com
       2 | Rubie77@yahoo.com
       2 | Rudolph93@gmail.com
       2 | Russ_Boyle@hotmail.com
       2 | Russell46@gmail.com
       2 | Ruth4@yahoo.com
       2 | Ruthe10@gmail.com
       2 | Ruthie_Brown@gmail.com
       2 | Ruthie_Lakin@hotmail.com
       2 | Ryan21@yahoo.com
       2 | Ryan92@hotmail.com
       2 | Rylan.Ondricka@yahoo.com
       2 | Sadie19@gmail.com
       2 | Sadye87@yahoo.com
       2 | Sally91@hotmail.com
       2 | Samanta36@yahoo.com
       2 | Samantha.Fisher@yahoo.com
       2 | Samson38@hotmail.com
       2 | Sandy16@yahoo.com
       2 | Santina27@yahoo.com
       2 | Santina43@gmail.com
       2 | Santino.Corwin@gmail.com
       2 | Sarah_Walker@gmail.com
       2 | Sasha52@hotmail.com
       2 | Saul.Reichert@yahoo.com
       2 | Schuyler.Weber@hotmail.com
       2 | Scot61@gmail.com
       2 | Scot83@yahoo.com
       2 | Scotty68@gmail.com
       2 | Scotty90@hotmail.com
       2 | Sebastian1@gmail.com
       2 | Sebastian23@yahoo.com
       2 | Sebastian52@hotmail.com
       2 | Sebastian54@hotmail.com
       2 | Sedrick33@gmail.com
       2 | Serena3@gmail.com
       2 | Shaina51@yahoo.com
       2 | Shanel28@gmail.com
       2 | Shanie_Russel@yahoo.com
       2 | Shaniya21@gmail.com
       2 | Shaniya47@yahoo.com
       2 | Shanna3@yahoo.com
       2 | Shannon4@yahoo.com
       2 | Shanon53@yahoo.com
       2 | Shea64@yahoo.com
       2 | Shemar16@yahoo.com
       2 | Shemar68@gmail.com
       2 | Shirley95@yahoo.com
       2 | Sibyl96@hotmail.com
       2 | Sienna.Hettinger@hotmail.com
       2 | Sigmund_Stark@hotmail.com
       2 | Sigmund.Fay@gmail.com
       2 | Sigmund15@gmail.com
       2 | Sigurd.Moen@yahoo.com
       2 | Silas69@yahoo.com
       2 | Simeon96@hotmail.com
       2 | Simone98@gmail.com
       2 | Sincere_Kemmer@gmail.com
       2 | Skye97@gmail.com
       2 | Solon61@yahoo.com
       2 | Sonny38@hotmail.com
       2 | Sonya52@hotmail.com
       2 | Sophie22@gmail.com
       2 | Stacey43@hotmail.com
       2 | Stefanie_Kutch@gmail.com
       2 | Stella80@gmail.com
       2 | Stella90@hotmail.com
       2 | Stephanie17@hotmail.com
       2 | Summer48@hotmail.com
       2 | Susie52@gmail.com
       2 | Sven.Bins@hotmail.com
       2 | Sven55@yahoo.com
       2 | Syble_Mayert@gmail.com
       2 | Sydni14@hotmail.com
       2 | Sydni69@gmail.com
       2 | Sydnie21@yahoo.com
       2 | Tamia.Denesik@yahoo.com
       2 | Tania83@yahoo.com
       2 | Tanner17@hotmail.com
       2 | Tanya48@gmail.com
       2 | Tanya6@gmail.com
       2 | Tara6@yahoo.com
       2 | Tatyana4@gmail.com
       2 | Tavares83@hotmail.com
       2 | Teagan66@yahoo.com
       2 | Terrell33@gmail.com
       2 | Terrill97@gmail.com
       2 | Terry56@yahoo.com
       2 | Thaddeus_Jacobson@hotmail.com
       2 | Theo17@gmail.com
       2 | Therese50@gmail.com
       2 | Thomas69@gmail.com
       2 | Tia_Stokes@yahoo.com
       2 | Tia72@yahoo.com
       2 | Tiara2@hotmail.com
       2 | Tiara34@gmail.com
       2 | Tierra34@hotmail.com
       2 | Timothy98@yahoo.com
       2 | Tobin63@gmail.com
       2 | Toby24@hotmail.com
       2 | Tom94@gmail.com
       2 | Tomas44@yahoo.com
       2 | Toney55@hotmail.com
       2 | Torey26@hotmail.com
       2 | Toy0@yahoo.com
       2 | Tracy92@gmail.com
       2 | Travon18@yahoo.com
       2 | Travon96@hotmail.com
       2 | Tre70@gmail.com
       2 | Tremayne87@hotmail.com
       2 | Trenton_Murphy@yahoo.com
       2 | Trenton40@yahoo.com
       2 | Tressie4@gmail.com
       2 | Tristian21@yahoo.com
       2 | Trudie85@gmail.com
       2 | Tyra8@hotmail.com
       2 | Tyrique47@gmail.com
       2 | Ubaldo.Stamm@gmail.com
       2 | Una.Thiel@yahoo.com
       2 | Una94@gmail.com
       2 | Ursula21@yahoo.com
       2 | Valentine96@hotmail.com
       2 | Vergie14@hotmail.com
       2 | Verla58@hotmail.com
       2 | Verlie1@yahoo.com
       2 | Verlie88@yahoo.com
       2 | Vernon72@yahoo.com
       2 | Veronica83@yahoo.com
       2 | Vickie40@yahoo.com
       2 | Vickie91@yahoo.com
       2 | Vicky47@hotmail.com
       2 | Victor17@hotmail.com
       2 | Victor83@yahoo.com
       2 | Vincenza1@yahoo.com
       2 | Violet20@hotmail.com
       2 | Virgie75@gmail.com
       2 | Virginia34@yahoo.com
       2 | Viviane33@yahoo.com
       2 | Vivianne0@gmail.com
       2 | Wade73@hotmail.com
       2 | Wade96@hotmail.com
       2 | Ward14@hotmail.com
       2 | Weldon30@gmail.com
       2 | Weldon36@hotmail.com
       2 | Wilburn98@yahoo.com
       2 | Wiley91@hotmail.com
       2 | Wilford_Keebler@hotmail.com
       2 | Wilford78@gmail.com
       2 | Wilfrid7@hotmail.com
       2 | Wilson28@yahoo.com
       2 | Wilton52@gmail.com
       2 | Wilton68@yahoo.com
       2 | Winston7@yahoo.com
       2 | Woodrow19@yahoo.com
       2 | Xander0@yahoo.com
       2 | Xavier52@gmail.com
       2 | Yolanda67@yahoo.com
       2 | Zachariah43@yahoo.com
       2 | Zachary48@yahoo.com
       2 | Zelda11@gmail.com
       2 | Zena78@gmail.com
       2 | Zion.Oberbrunner@gmail.com
       2 | Zion62@hotmail.com
       2 | Zoe66@yahoo.com
       2 | Zoey6@hotmail.com
       2 | Zoila5@hotmail.com
       2 | Zoila72@gmail.com
```

**Comment by JasonCust** [comment]
@Marak Is this still live in 4.1?
```javascript
> const faker = require('faker');
undefined
> faker.unique
undefined
```

**Comment by Marak** [comment]
@JasonCust 

Try cloning the `master` branch from Github, it should be in there.

There is much improvement for uniqueness, especially around email APIs.

**Comment by JasonCust** [comment]
@Marak Works. Any idea of when it will be published to NPM?

**Comment by hansede** [comment]
This problem is happening because of this line:
`switch (faker.random.number(2)) {`

You are choosing from each email generation scheme with equal probability, but that means that you are not assigning equal probability to every possible outcome because some schemes generate far more possible outcomes than others. 

You could dramatically decrease the odds of a collision by removing cases 0 and 1 and always generating from case 2. From there you could further dramatically decrease collisions by appending `faker.random.number(99999)` Instead of `faker.random.number(99)`.

**Comment by Marak** [comment]
@hansede -

That actually makes a lot of sense. Thank you for taking the time to explain the issue.

I didn't personally write that switch case statement, but it's now obvious to me why we are seeing the high rate of collisions for this method. I'm going to rewrite the email generation API based on your suggestions. Should be easy.

To be clear, the addition of `faker.unique` is still a very good general solution across `faker` for ensuring uniqueness. 

Fixing this newly identified issue specifically with email generation will be a simple and good fix for ensuring "more random" email addresses.

**Comment by pkuczynski** [comment]
@Marak, any plans to release a new version with `faker.unique`?

**Comment by pkuczynski** [comment]
@Marak ping

**Comment by fr3fou** [comment]
any news?

**Comment by rsaphala** [comment]
any updates on this?

**Comment by liquidvisual** [comment]
Any updates?

**Comment by malimccalla** [comment]
Would love to see this released soon

**Comment by tarikhamilton** [comment]
What are people doing as a workaround? Kinda shocked this hasn't been released when it's been supposedly complete for a year.

**Comment by Marak** [comment]
@tarikhamilton - I've deleted your comment and temporarily removed your ability to further contribute. 

Next time please try to be more civil in your tone and comments.

**Comment by AbdealiJK** [comment]
Seems like we're having issues with uniqueness too - and came across this thread.
Is there a release in the works that we should be waiting for - or is the recommended approach to leverage `faker.unique` to use the github master branch directly ?

Maybe having a `rc` or `beta` version on `npm` registry would help users who want to test it out as this seems to be a very wanted feature :)

**Comment by Zoha** [comment]
In my opinion, the most important reason for using a faker is to seeding the database. And what is the benefit when the data is not unique?
This Librarian still can't create unique data (in loop) ?

**Comment by berniedurfee** [comment]
Bump 👍

**Comment by abettke** [comment]
Bumping here with enthusiastic interest

**Comment by Ben-Kincaid** [comment]
Also very interested in this.. currently using from master branch directly and `.unique` works great. But would love to see this in the main package, as `@types/faker` doesn't currently have types for `.unique`.

**Comment by elseify-r-medicalib** [comment]
Does anybody know how to use faker unique with faker.fake generator ?

**Comment by jasonlimantoro** [comment]
Should add this in the docs:

for example, if you want to generate random words using the `faker.random.words` API,

```ts
// first argument is the method, the second argument is the parameters (as an array) passed to the method.
const word = faker.unique(faker.random.words, [3], { maxRetries: 3 });
```

**Comment by rdsedmundo** [comment]
IHMO this `faker.unique` thing should be applied by default in the library for all the methods. I don't think anyone expects collisions by default.

**Comment by jasonlimantoro** [comment]
I agree with @rdsedmundo, it sounds a bit stupid to have to explicitly implement the uniqueness. I guess for now we have to declare our own wrapper function until this issue got more attention sometime in the future.

**Comment by rdsedmundo** [comment]
By the way, I'm still sending duplicated stuff. Jest runs different test files in parallel in different child processes and the [package internal memory](https://github.com/Marak/faker.js/blob/master/vendor/unique.js#L8) is not shared.

**Comment by jasonlimantoro** [comment]
Argh, that's nasty. No idea how to work around that issue other than running the jest tests sequentially (using the `--runInBand` flag).

**Comment by rdsedmundo** [comment]
I think the problem is here: 
https://github.com/Marak/faker.js/blob/c9764abd20a198e318367952a94587eda1bac645/lib/mersenne.js#L5

I don't think using the date as a seed is the best approach. I added the following call, and so far I wasn't able to get any duplicates, I'll report back if it wasn't sufficient:
```js
faker.seed(Math.floor(1_000_000_000 * Math.random()));
```

With this, I don't even need the `faker.unique` wrapper.

**Comment by jasonlimantoro** [comment]
Did you initialize the seed on every test file?

**Comment by rdsedmundo** [comment]
No, just once.

**faker.ts**
```ts
import faker from 'faker';

faker.seed(Math.floor(1_000_000_000 * Math.random()));

export { faker };
```

Then I import this file instead of importing from `faker` directly.

**Comment by maximkott** [comment]
I've had this problem a while ago and solved it in a similar way. The main idea is - most of the time I use faker during tests, if using jest, one can simply do:

```js
beforeAll(() => faker.seed(Math.floor(1_000_000_000 * Math.random())))
```

This will change the seed once for every test suite.

---

Also, please don't add any **additional** arguments to the **faker.unique** method. Something like this is **not** needed:

```js
faker.unique(faker.random.words, [3])
```

Instead users should do:

```js
faker.unique(() => faker.random.words(3))
```
